db/compaction_picker.cc

```diff
@@ -52,9 +54,11 @@ bool FindIntraL0Compaction(const std::vector<FileMetaData*>& level_files,
   size_t new_compact_bytes_per_del_file = 0;
   for (span_len = 1; span_len < level_files.size(); ++span_len) {
     compact_bytes += static_cast<size_t>(level_files[span_len]->fd.file_size);
+    compensated_compact_bytes += level_files[span_len]->compensated_file_size;
     new_compact_bytes_per_del_file = compact_bytes / span_len;
     if (level_files[span_len]->being_compacted ||
-        new_compact_bytes_per_del_file > compact_bytes_per_del_file) {
+        new_compact_bytes_per_del_file > compact_bytes_per_del_file ||
+        compensated_compact_bytes > max_compaction_bytes) {
```

Comment:
I've been thinking about the problem with L0->L1 compactions recently (really L0->Lbase). In addition to the size of those compactions, they are problematic because an L0->L1 compaction "locks" L1, preventing a concurrent compaction of L1->L2. The idea I've been tossing around is to allow flushes to create more than 1 sstable in L0, using the existing compaction logic to create table boundaries based on the "grandparent" table boundaries in L1. In order to make this work, L0 would have to be structured as a series of sublevels where each sublevel can contain 1 or more sstables. Not impossible, but somewhat of a heavy lift.

Passing this along in the hopes that it will spark a better idea.

Reply:
CC @maysamyabandeh who plans to combine leveled compaction with universal. This idea can be a subset of the combined compaction.

Actually, if we only consider L1, it is relatively easier to implement, because with dynamic leveling, usually are empty levels between the level L0 is merging into and "L1". If "L1" is being occupied, we can compact L0 files to a new level before "L1". Whether we avoid the problem of compaction being too large for L1s to L2 is another question though.